File: trail/src/main/java/com/amalbit/trail/marker/MarkerOverlayView.java
package com.amalbit.trail.marker;

import android.content.Context;
import android.graphics.Canvas;
import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.constraintlayout.widget.ConstraintSet;

import android.util.AttributeSet;
import android.view.View;

import com.amalbit.trail.marker.OverlayMarker.MarkerRemoveListner;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.Projection;

import java.util.ArrayList;
import java.util.List;

public class MarkerOverlayView extends ConstraintLayout implements MarkerRemoveListner {

//    private final Object mSvgLock = new Object();

    /**
     * TO be converted to a HashMap.
     **/
    private List<OverlayMarker> overlayMarkers;

    public MarkerOverlayView(Context context) {
        super(context);
        init();
    }

    public MarkerOverlayView(Context context,
                             @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    private void init() {
        setLayerType(View.LAYER_TYPE_SOFTWARE, null);
        overlayMarkers = new ArrayList<>();
    }

    public void addMarker(OverlayMarker overlayMarker, Projection projection) {
        overlayMarker.setScreenPoint(projection.toScreenLocation(overlayMarker.getLatLng()));
        overlayMarker.setMarkerRemoveListner(this);
        overlayMarkers.add(overlayMarker);
        overlayMarker.anchorView2 = new AnchorView2(getContext());
        addMarkerToLayout(overlayMarker);
    }

    public OverlayMarker findMarkerById(int markerId) {
        for (OverlayMarker marker : overlayMarkers) {
            if (marker.getMarkerId() == markerId) {
                return marker;
            }
        }
        return null;
    }

    @Override
    public void onRemove(OverlayMarker overlayMarker) {
        overlayMarkers.remove(overlayMarker);
        invalidate();
    }

    public void onCameraMove(GoogleMap googleMap) {
        for (OverlayMarker overlayMarker : overlayMarkers) {
            overlayMarker.setScreenPoint(googleMap.getProjection().toScreenLocation(overlayMarker.getLatLng()));
            overlayMarker.setMapBearing(googleMap.getCameraPosition().bearing);
            updateMarkerPosition(overlayMarker);
        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
    }


    private void updateMarkerPosition(OverlayMarker overlayMarker) {
        overlayMarker.anchorView2.setAbsolutePosition(overlayMarker.getScreenPoint().x, overlayMarker.getScreenPoint().y);
    }

    private void addMarkerToLayout(OverlayMarker overlayMarker) {
        addView(overlayMarker.anchorView2, 1, 1);
        ConstraintSet set = new ConstraintSet();
        set.clone(this);
        set.connect(overlayMarker.anchorView2.getId(), ConstraintSet.TOP, ConstraintSet.PARENT_ID, ConstraintSet.TOP, 0);
        set.connect(overlayMarker.anchorView2.getId(), ConstraintSet.LEFT, ConstraintSet.PARENT_ID, ConstraintSet.LEFT, 0);
        set.applyTo(this);
        updateMarkerPosition(overlayMarker);
    }
}

File: trail/src/main/java/com/amalbit/trail/marker/MarkerOverlayView.kt
package com.amalbit.trail.marker

import android.content.Context
import android.util.AttributeSet
import android.view.Gravity
import android.view.ViewGroup
import android.widget.Button
import android.widget.FrameLayout
import com.amalbit.trail.marker.MarkerHolder.MarkerRemoveListner
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.Projection
import java.util.*

class MarkerOverlayView @JvmOverloads constructor(
        context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
) : FrameLayout(context, attrs, defStyleAttr), MarkerRemoveListner {
    private var markerHolders: MutableList<MarkerHolder> = ArrayList()

    init {
        setLayerType(LAYER_TYPE_HARDWARE, null)
    }

    fun addMarker(markerHolder: MarkerHolder, projection: Projection) {
        markerHolder.screenPoint = projection.toScreenLocation(markerHolder.latLng)
        markerHolder.markerRemoveListner = this
        markerHolders.add(markerHolder)
        markerHolder.view = Button(context)
        addMarkerToLayout(markerHolder)
    }

    fun findMarkerById(markerId: Int): MarkerHolder? {
        for (marker in markerHolders) {
            if (marker.markerId == markerId) {
                return marker
            }
        }
        return null
    }

    fun onCameraMove(googleMap: GoogleMap) {
        for (overlayMarker in markerHolders) {
            overlayMarker.screenPoint = googleMap.projection.toScreenLocation(overlayMarker.latLng)
            overlayMarker.mapBearing = googleMap.cameraPosition.bearing
            updateMarkerPosition(overlayMarker)
        }
        requestLayout()
    }

    private fun updateMarkerPosition(markerHolder: MarkerHolder) {
        val dx = convertX(markerHolder.screenPoint!!.x.toFloat()).toInt()
        val dy = convertY(markerHolder.screenPoint!!.y.toFloat()).toInt()

        (markerHolder.view?.layoutParams as LayoutParams).run {
            setMargins(dx, dy, 0, 0)
        }
    }

    private fun addMarkerToLayout(markerHolder: MarkerHolder) {
        val layoutParams = LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        layoutParams.gravity = Gravity.CENTER
        addView(markerHolder.view, layoutParams)
        updateMarkerPosition(markerHolder)
    }

    private fun convertX(x: Float): Float {
        return x - width / 2f
    }

    private fun convertY(y: Float): Float {
        return y - height / 2f
    }

    override fun onRemove(markerHolder: MarkerHolder?) {
        markerHolders.remove(markerHolder)
        invalidate()
    }
}